crates/sui-storage/src/indexes.rs

```diff
@@ -1022,4 +1102,437 @@ impl IndexStore {
+                .map(|(_coin_type, obj_id, coin)| (coin_type_str.clone(), obj_id, coin));
+
+        let mut balance = 0u128;
+        let mut num_coins = 0;
+        for (_coin_type, _obj_id, coin_info) in coins {
+            balance += coin_info.balance as u128;
+            num_coins += 1;
+        }
+
+        Ok(TotalBalance { balance, num_coins })
+    }
+
+    /// Read all balances for a `SuiAddress` from the backend database
+    pub fn get_all_balances_from_db(
+        metrics: Arc<IndexStoreMetrics>,
+        coin_index: DBMap<CoinIndexKey, CoinInfo>,
+        owner: SuiAddress,
+    ) -> SuiResult<Arc<HashMap<TypeTag, TotalBalance>>> {
+        metrics.all_balance_lookup_from_db.inc();
+        let mut balances: HashMap<TypeTag, TotalBalance> = HashMap::new();
+        let coins = Self::get_owned_coins_iterator(&coin_index, owner, None)?
+            .map(|(coin_type, obj_id, coin)| (coin_type, obj_id, coin))
+            .group_by(|(coin_type, _obj_id, _coin)| coin_type.clone());
+        for (coin_type, coins) in &coins {
+            let mut total_balance = 0u128;
+            let mut coin_object_count = 0;
+            for (_coin_type, _obj_id, coin_info) in coins {
+                total_balance += coin_info.balance as u128;
+                coin_object_count += 1;
+            }
+            let coin_type =
+                TypeTag::Struct(Box::new(parse_sui_struct_tag(&coin_type).map_err(|e| {
+                    SuiError::ExecutionError(format!(
+                        "Failed to parse event sender address: {:?}",
+                        e
+                    ))
+                })?));
+            balances.insert(
+                coin_type,
+                TotalBalance {
+                    num_coins: coin_object_count,
+                    balance: total_balance,
+                },
+            );
+        }
+        Ok(Arc::new(balances))
+    }
+
+    async fn invalidate_deleted_coins(&self, owners: &[OwnerIndexKey]) -> SuiResult {
+        for owner in owners.iter() {
+            // This coin should be in the index if it is getting deleted
+            let object_info = self.tables.owner_index.get(owner)?;
+            if let Some(object_info) = object_info {
+                if let Ok(type_tags) = match object_info.type_ {
+                    ObjectType::Package => Err(anyhow!("Cannot create StructTag from Package")),
+                    ObjectType::Struct(move_object_type) => Ok(move_object_type.type_params()),
+                } {
+                    let type_tag = if let Some(type_tag) = type_tags.first() {
+                        type_tag.clone()
+                    } else {
+                        TypeTag::Struct(Box::new(GAS::type_()))
+                    };
+                    let _rwlock = self
+                        .caches
+                        .per_coin_rw_locks
+                        .acquire_lock((owner.0, type_tag.clone()))
```

Comment:
I think it's helpful to merge `invalidate_deleted_coins` and `invalidate_added_coins` so only need to acquire the write lock once for `per_coin_rw_locks`